core/authors/views.py

```diff
@@ -186,4 +191,135 @@ def check_is_following(self, request, pk):
         
         return Response({
             "isFollowingUser": follow.exists()
-        }, status=200)
\ No newline at end of file
+        }, status=200)
+
+    # All posts the currently auth'd user can see of pk
+    # /author/{AUTHOR_ID}/posts
+    @action(detail=True, url_path="posts")
+    def posts(self, request, pk=None):
+        page = int(request.query_params.get("page", 0)) + 1 # Must offset page by 1
+        if page < 1:
+            return Response("Page number must be positive", status=400)
+
+        # TODO: size should be limited?
+        size = int(request.query_params.get("size", 50))
+        if size < 0:
+            return Response("Size must be positive", status=400)
+
+        if not pk:
+            # TODO should it be a text response?
+            return Response("You must specify an author.",status=400)
+
+        # Only return public posts if the user isn't authenticated
+        if request.user.is_anonymous:
+            posts = Posts.objects.all().filter(author=pk, visibility__in=["PUBLIC", "SERVERONLY"])
+        # else if is other_server:
+        #     posts = Posts.objects.all().filter(author=pk, visibility__in=["PUBLIC"])
+        else:
+            requestingAuthor = request.user.author.id # Should be guaranteed because we checked above
+
+            # post_types will track what level of posts a user can see
+            post_types = ["PUBLIC", "SERVERONLY"]
+            # convert to dict for dat O(1)
+            # Note: this is terrible, we should be using the database more directly
+            requesterFriends = {}
+            for friend in get_friends_from_pk(requestingAuthor):
+                requesterFriends[friend] = True
+
+            # Check if they are direct friends
+            if requesterFriends.get(pk, False):
+                post_types += ["FRIENDS", "FOAF"]
+            else: # They are not direct friends, so we should check if they share any friends
+                for friend in get_friends_from_pk(pk):
+                    if requesterFriends.get(friend, False):
+                        post_types += ["FOAF"]
+                        break # we don't need to check any more friends
+
+            try:
+                posts = Posts.objects.all().filter(author=pk, visibility__in=post_types)
+                # TODO: requestingAuthor is the one it should be visibleTo
+                posts |= Posts.objects.all().filter(author=pk, visibility="PRIVATE", visibleTo__contains=get_author_url(str(requestingAuthor)))
```

Comment:
Shouldn't this be visibleTo__contains=[get_author_url(str(requestingAuthor))]?

Reply:
should be correct now